File: android/src/main/java/com/thunderstorm/app/android/viewmodel/SetupViewModel.kt
package com.thunderstorm.app.android.viewmodel

import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.ui.text.input.TextFieldValue
import androidx.lifecycle.ViewModel
import com.thunderstorm.app.model.SearchCityResult
import com.thunderstorm.app.networking.NetworkingClient
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch

class SetupViewModel : ViewModel() {

    val searchFieldValue = mutableStateOf(TextFieldValue(""))
    val selectedCity = mutableStateOf<SearchCityResult?>(null)
    val cityAutocompleteItems = mutableStateOf(mutableListOf<SearchCityResult>())
    val allowNavigateNext = mutableStateOf(false)

    private val asyncScope = CoroutineScope(Dispatchers.IO + Job())

    fun fetchCitiesForSearch() {
        asyncScope.launch {
            NetworkingClient().getMatchingCitiesForSearch(
                query = searchFieldValue.value.text,
                onResultAvailable = { result ->
                    cityAutocompleteItems.value.clear()
                    result.forEach { searchCityResult ->
                        cityAutocompleteItems.value.add(searchCityResult)
                    }
                }
            )
        }
    }

}

File: ios/ios/ViewModels/SetupViewModel.swift
import SwiftUI
import Foundation
import shared

class SetupViewModel: ObservableObject {

    @Published var cityList: [SearchCityResult] = []
    @Published var allowNavigateNext: Bool = false
    @Published var selectedCity: SearchCityResult? = nil
    
    @Published var selectionTemperature = 0
    @Published var selectionSpeed = 0
    @Published var selectionPrecip = 0
    @Published var selectionAir = 0
    
    func fetchCitiesForSearch(query: String?) {
        if (query != "" && query != nil) {
            NetworkingClient().getMatchingCitiesForSearch(
                query: query,
                onResultAvailable: { result in
                    self.cityList.removeAll()
                    if (!result.isEmpty) {
                        result.forEach { item in
                            self.cityList.append(item)
                        }
                    }
                }
            )
        }
    }
    
}
